xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,373 @@
+        (double) config.maxRingSize);
+    List<RingEntry> ring = new ArrayList<>();
+    double currentHashes = 0.0;
+    double targetHashes = 0.0;
+    for (Map.Entry<EquivalentAddressGroup, Long> entry : serverWeights.entrySet()) {
+      EquivalentAddressGroup addrKey = entry.getKey();
+      double normalizedWeight = (double) entry.getValue() / totalWeight;
+      // TODO(chengyuanzhang): is using the list of socket address correct?
+      StringBuilder sb = new StringBuilder(addrKey.getAddresses().toString());
+      sb.append('_');
+      targetHashes += scale * normalizedWeight;
+      long i = 0L;
+      while (currentHashes < targetHashes) {
+        sb.append(i);
+        long hash = hashFunc.hashAsciiString(sb.toString());
+        ring.add(new RingEntry(hash, addrKey));
+        i++;
+        currentHashes++;
+        sb.deleteCharAt(sb.length() - 1);
+      }
+    }
+    Collections.sort(ring);
+
+    for (EquivalentAddressGroup addr : addedAddrs) {
+      Attributes attr = Attributes.newBuilder().set(
+          STATE_INFO, new AtomicReference<>(ConnectivityStateInfo.forNonError(IDLE))).build();
+      final Subchannel subchannel = helper.createSubchannel(
+          CreateSubchannelArgs.newBuilder()
+              .setAddresses(latestAddrs.get(addr)).setAttributes(attr).build());
+      subchannel.start(new SubchannelStateListener() {
+        @Override
+        public void onSubchannelState(ConnectivityStateInfo newState) {
+          processSubchannelState(subchannel, newState);
+        }
+      });
+      subchannels.put(addr, subchannel);
+    }
+    List<Subchannel> removedSubchannels = new ArrayList<>();
+    for (EquivalentAddressGroup addr : removedAddrs) {
+      removedSubchannels.add(subchannels.remove(addr));
+    }
+
+    // Update the picker before shutting down the subchannels, to reduce the chance of race
+    // between picking a subchannel and shutting it down.
+    updateBalancingState(new RingHashPicker(
+        Collections.unmodifiableList(ring), ImmutableMap.copyOf(subchannels)));
+    for (Subchannel subchann : removedSubchannels) {
+      shutdownSubchannel(subchann);
+    }
+  }
+
+  private static void shutdownSubchannel(Subchannel subchannel) {
+    subchannel.shutdown();
+    getSubchannelStateInfoRef(subchannel).set(ConnectivityStateInfo.forNonError(SHUTDOWN));
+  }
+
+  /**
+   * Converts list of {@link EquivalentAddressGroup} to {@link EquivalentAddressGroup} set and
+   * remove all attributes. The values are the original EAGs.
+   */
+  private static Map<EquivalentAddressGroup, EquivalentAddressGroup> stripAttrs(
+      List<EquivalentAddressGroup> groupList) {
+    Map<EquivalentAddressGroup, EquivalentAddressGroup> addrs = new HashMap<>(groupList.size() * 2);
+    for (EquivalentAddressGroup group : groupList) {
+      addrs.put(stripAttrs(group), group);
+    }
+    return addrs;
+  }
+
+  private static EquivalentAddressGroup stripAttrs(EquivalentAddressGroup eag) {
```

Comment:
I don't get it, why return a list?

> It's hard to tell a collection of EAGs like Set<EquivalentAddressGroup> removedAddrs is stripped or not at a glance.

This is true, readers need to follow the logic to see where these `EquivalentAddressGroup`s come from. Most of the time, it should be using the attribute-stripped address (if not, it can be changed to) for main logics. The original EAGs are needed only for creating subchannels.

Reply:
I mean return List<SocketAddress>.  Because EAG stripping attributes is just `eag.getAddresses()`, why bother wrap it to EAG again.

Reply:
Well, there is no difference. This wrapping is cheap. While with `List<SocketAddress>`, it ends up with always carrying the list for things like `Set<List<SocketAddress>>`, which sacrifices simplicity and cleanness. The even worse thing is that it is throwing away encapsulation by unnecessarily exposing `SocketAddress`. The load balancer logic doesn't need to deal with anything at `SocketAddress` level.